driver/src/orderbook/streamed/state.rs

```diff
@@ -323,17 +323,122 @@ mod tests {
+                valid_until: $valid_until,
+                price_numerator: $sell_amount,
+                price_denominator: $buy_amount,
+            };
+            $state = $state
+                .apply_event(&Event::OrderPlacement(event), $batch)
+                .unwrap();
+        };
+        (to $state:ident in batch $batch:expr, trade order number $index:expr, from user $user:expr, selling $sell_amount:expr, for $buy_amount:expr) => {
+            let event = Trade {
+                owner: address($user),
+                order_id: $index,
+                executed_sell_amount: $sell_amount,
+                executed_buy_amount: $buy_amount,
+                ..Default::default()
+            };
+            $state = $state.apply_event(&Event::Trade(event), $batch).unwrap();
+        };
+        (to $state:ident in batch $batch:expr, revert trade order number $index:expr, from user $user:expr, selling $sell_amount:expr, for $buy_amount:expr) => {
+            let event = TradeReversion {
+                owner: address($user),
+                order_id: $index,
+                executed_sell_amount: $sell_amount,
+                executed_buy_amount: $buy_amount,
+                ..Default::default()
+            };
+            $state = $state
+                .apply_event(&Event::TradeReversion(event), $batch)
+                .unwrap();
+        };
+        (to $state:ident in batch $batch:expr, submit solution from user $user:expr, with fee $fee:expr) => {
+            let event = SolutionSubmission {
+                submitter: address($user),
+                burnt_fees: U256::from($fee),
+                ..Default::default()
+            };
+            $state = $state
+                .apply_event(&Event::SolutionSubmission(event), $batch)
+                .unwrap();
+        };
+    }
+
+    macro_rules! assert_balance {
+        (in $state:ident batch $batch:expr, user $user:expr, has token $token:expr, balance $balance:expr) => {
+            assert_eq!(
+                account_state(&$state, $batch).read_balance($token, address($user)),
+                U256::from($balance)
+            );
+        };
+    }
+
+    macro_rules! assert_used_amount {
+        (in $state:ident batch $batch:expr, of order number $index:expr, from user $user:expr, is $amount:expr) => {
+            assert_eq!(
+                $state
+                    .orders
+                    .get(&(address($user), $index))
+                    .unwrap()
+                    .get_used_amount($batch),
+                $amount
+            );
+        };
+    }
+
     fn address(n: u64) -> Address {
         Address::from_low_u64_be(n)
     }
 
     fn state_with_fee() -> State {
         let mut state = State::default();
-        let event = TokenListing {
-            token: address(0),
-            id: 0,
-        };
-        state = state.apply_event(&Event::TokenListing(event), 0).unwrap();
+        apply_event!(to state in batch 0, list token 0);
```

Comment:
This is pretty strange to look at. I think mostly because `list token` is two separate words and the word `to` and `in` are they even anything?

Reply:
This is strange if you're used to functions because it is a macro. I intentionally tried to pick something that reads more like a human language. The meaning of the words come entirely from the macro so the syntax here can be nearly anythign.